code/modules/projectiles/projectile.dm

```diff
@@ -631,4 +631,58 @@ var/list/impact_master = list()
 
 /obj/item/projectile/attack_hand(mob/user)
 	if(timestopped)
-		..()
\ No newline at end of file
+		..()
+
+/obj/item/projectile/friendlyCheck
+	invisibility = 101
+	rotate = 0
+	damage = 0
+	nodamage = 1
+	var/atom/impact = null
+
+/obj/item/projectile/friendlyCheck/process()
+	OnFired()
+	while(!impact && loc && (kill_count > 0))
+		if(dist_x > dist_y)
+			bresenham_step(dist_x,dist_y,dx,dy)
+		else
+			bresenham_step(dist_y,dist_x,dy,dx)
+	return impact
+
+/obj/item/projectile/friendlyCheck/bresenham_step(var/distA, var/distB, var/dA, var/dB)
+	kill_count--
+	total_steps++
+	if(error < 0)
+		var/atom/step = get_step(src, dB)
+		if(!step)
+			bullet_die()
+		src.Move(step)
+		error += distA
+		bump_original_check()
+	else
+		var/atom/step = get_step(src, dA)
+		if(!step)
+			bullet_die()
+		src.Move(step)
+		error -= distB
+		dir = dA
+		if(error < 0)
+			dir = dA + dB
+		bump_original_check()
+
+/obj/item/projectile/friendlyCheck/Bump(var/atom/A)
+	if(bumped)	return 0
+
+	bumped = 1
+
+	if(istype(A, /obj/structure/bed/chair/vehicle))
```

Comment:
As much as I would love to push for an OOP approach, it will be out of scope for how small the PR actually is, which is not something I like to say.

You can just put a proc in projectile called get_hit_atom(var/atom/a), put the current code into there, then change the current code to impact = get_hit_atom(whateverthisvariscalled) and we'll think about a better solution for a followup systems PR.